modules/images/regenerate-existing-images/background-process/class-perflab-background-process.php

```diff
@@ -0,0 +1,236 @@
+	 *
+	 * @since n.e.x.t
+	 *
+	 * @var string Background process action name.
+	 */
+	const BG_PROCESS_ACTION = 'background_process_handle_request';
+
+	/**
+	 * Job instance.
+	 *
+	 * @since n.e.x.t
+	 *
+	 * @var Perflab_Background_Job Background job instance.
+	 */
+	private $job;
+
+	/**
+	 * Perflab_Background_Process constructor.
+	 *
+	 * Adds the necessary hooks to trigger process handling.
+	 *
+	 * @since n.e.x.t
+	 *
+	 * @return void
+	 */
+	public function __construct() {
+		add_action( 'wp_ajax_' . Perflab_Background_Process::BG_PROCESS_ACTION, array( $this, 'handle_request' ) );
+		add_action( 'wp_ajax_nopriv_' . Perflab_Background_Process::BG_PROCESS_ACTION, array( $this, 'handle_request' ) );
+	}
+
+	/**
+	 * Handle incoming request to run the batch of job.
+	 *
+	 * @since n.e.x.t
+	 *
+	 * @throws Exception Invalid request for background process.
+	 *
+	 * @return void
+	 */
+	public function handle_request() {
+		try {
+			// Exit if nonce varification fails.
+			$nonce_check = check_ajax_referer( self::BG_PROCESS_ACTION, 'nonce', false );
+
+			// If nonce check fails, fallback to key checking.
+			if ( false === $nonce_check ) {
+				throw new Exception( __( 'Invalid nonce passed to process.', 'performance-lab' ) );
+			}
+
+			$job_id = isset( $_REQUEST['job_id'] ) ? absint( sanitize_text_field( $_REQUEST['job_id'] ) ) : 0;
+
+			// Job ID is mandatory to be specified.
+			if ( empty( $job_id ) ) {
+				throw new Exception( __( 'No job specified to execute.', 'performance-lab' ) );
+			}
+
+			$this->job = new Perflab_Background_Job( $job_id );
+
+			// Silently exit if the job is not ready to run.
+			if ( ! $this->job->should_run() ) {
+				return;
+			}
+
+			$this->job->lock(); // Lock the process for this job before running.
+			$this->run(); // Run the job.
+
+			// Once job ran successfully, change its status to queued.
+			$this->job->set_status( Perflab_Background_Job::JOB_STATUS_PARTIAL );
+
+		} catch ( Exception $e ) {
+			if ( $this->job instanceof Perflab_Background_Job ) {
+				$error = new WP_Error( 'perflab_job_failure', $e->getMessage() );
+				$this->job->set_error( $error );
+			}
+		} finally {
+			if ( $this->job instanceof Perflab_Background_Job ) {
+				// Unlock the process once everything is done.
+				$this->job->unlock();
+				$this->next_batch();
+			}
+		}
+	}
+
+	/**
+	 * Runs the process over a batch of job.
+	 *
+	 * As of now, it won't fetch the next batch if memory or time
+	 * is not exceeded, but this can be introduced later.
+	 *
+	 * Consumer code can fine-tune the batch size according to requirement.
+	 *
+	 * @since n.e.x.t
+	 *
+	 * @return void
+	 */
+	private function run() {
+		do {
+			/**
+			 * Consumer code will hook to this action to perform necessary tasks.
+			 *
+			 * @param array $data Job data.
+			 */
+			do_action( 'perflab_job_' . $this->job->get_name(), $this->job->get_data() );
+		} while ( ! $this->memory_exceeded() && ! $this->time_exceeded() );
```

Comment:
@ankitrox we should also check if the job is marked as complete here too.

```suggestion
		} while ( ! $this->memory_exceeded() && ! $this->time_exceeded() && ! $this->job->is_completed() );
```